example/notify_inval_entry.c

```diff
@@ -258,8 +268,22 @@ static void* update_fs_loop(void *data) {
         update_fs();
         if (!options.no_notify && lookup_cnt) {
             if(options.only_expire) {
-                assert(fuse_lowlevel_notify_expire_entry
-                   (se, FUSE_ROOT_ID, old_name, strlen(old_name), FUSE_LL_EXPIRE_ONLY) == 0);
+                int ret = fuse_lowlevel_notify_expire_entry
+                   (se, FUSE_ROOT_ID, old_name, strlen(old_name));
+                if (ret == -ENOSYS) {
+                    printf("fuse_lowlevel_notify_expire_entry not supported by kernel\n");
+                    printf("Next call to mountpoint will kill it...\n");
+                    raise(SIGINT);
+                    return NULL;
+                }
+                // forget is not being called
+                // the dentry is running just into a timeout
+                if (old_lookup_cnt == lookup_cnt) {
+                    assert(ret == -ENOENT);
+                } else {
+                    assert(ret == 0);
+                    old_lookup_cnt = lookup_cnt;
+                }
```

Comment:
What's the purpose of this code? It looks like we are mixing a unit test into an example, but I also do not understand what exactly is being tested here. It seems we are assuming that we know exactly what file system clients will do in-between iterations of this loop?

My feeling is that we should remove this entirely. It's not libfuses job to test if kernel functionality is implemented  correctly, especially if we can't even do so reliable. Let's just make sure that we test the userspace codepath.

Reply:
we talked about it before in the other PR ( https://github.com/libfuse/libfuse/pull/771#issuecomment-1502966224 and https://github.com/libfuse/libfuse/pull/771#issuecomment-1503191654 )

this examples creates a new file with the given timestamp each second. as `lookup_cnt` is not decreased if `expire` is used, the loop is always executed. so one needs to check if there is actually a file to be expired or expired is executed on a file that does not exist.

Reply:
we talked about this in PR #771 ( https://github.com/libfuse/libfuse/pull/771#issuecomment-1502966224 and https://github.com/libfuse/libfuse/pull/771#issuecomment-1503191654 )

a file named with the current timestamp is created every 1 sec. As `expire` does not reduce the `lookup_cnt` the if inside the loop is always executed. so one needs to destinguish if `expire` is successfully executed on a file that exists or on something that does not exist anymore.

Reply:
Sorry, I still do not understand what's being tested here.

Are you trying to test that the *kernel* correctly processes expire requests? If so, then we should check that a client stat() after expire() results in the right callback to FUSE, and that an invalidate request results in a callback to forget(). The current code seems to do neither. However, I would argue that it's not libfuse's job to test if the kernel implementation is correct anyway, especially if we can't do it reliably (tests may fail if there are other requests to the mountpoint). So I would drop this completely (unless I'm missing something crucial?).

Reply:
no we do not check if the expire request is handled correctly, only that it returns the correct return value.

because of
https://github.com/libfuse/libfuse/blob/8b77a82d3fc957a4662f9deb0d6b569428a955f1/example/notify_inval_entry.c#L269
we need to check the `expire` is executed because: 
1) "yes we accessed the file and want to expire it "(returns 0) 
or
2) "oh it's still an old one and does not exist anymore" (returns  -ENOENT)

if we do not want the `tfs_init` we also want to exit the program if no kernel support is available and `-ENOSYS` is returned

Reply:
As you point out yourself, fuse_session_exit() can be used. I'm personally not too happy about thee lookup_cnt checks - this is fragile and not under libfuse control. Kernel side is free to release cached inodes whenever it wants - will happen under memory pressure and on echo 3 >/proc/sys/vm/drop_caches. I.e. the unit test might spuriously fail.

Reply:
hi @bsbernd 
thanks for your feedback. you are correct with your arguments but i also think for this example those edge cases are a bit of an overkill. this example only creates/modifies a single (empty) file in a given update interval (default 1 sec) + you can set the timeout of the kernel caches (default 5 sec).

i guess if you really want to protect against drop_caches you need to change this line
https://github.com/libfuse/libfuse/blob/8b77a82d3fc957a4662f9deb0d6b569428a955f1/example/notify_inval_entry.c#L281-L283

to something like
```c++
if (old_lookup_cnt >= lookup_cnt) {
  old_lookup_cnt = lookup_cnt;
  assert(ret == -ENOENT);
} 
```

Reply:
I think this code should just be removed. As Bernd point out, it's fragile (that's a pain when running a CI), and secondly I still believe that this tests the kernel rather than libfuse (the return value comes from the kernel).

Reply:
i dont see how we can remove it. maybe we can add some locking around the manipulation of the counter.

when we `expire` `forget()` is not called. therefore the counter is not decreased. as the counter is not decreased the `if` within the `while`-loop is always called and as such always `expire`  is called: on existing or old non-existing files. only because of the counter we know if we should expect and error or a successful execution of `expire`.

if you have any better idea or if i mistunderstand something, please let me know

Reply:
This is kind of a test for libfuse + kernel and kerne sidel is not entirely returning what is actually does - this code is hacking it around and that certainly works most of the time.